packages/trail-core/lib/index.js

```diff
@@ -87,16 +88,35 @@ class TrailsManager {
           ("when" >= ${from.toISO()} AND "when" <= ${to.toISO()})
     `
 
+    // Prepare the query to get Count
+    const sqlCount = SQL`
+      SELECT
+          count(*)
+        FROM trails
+        WHERE
+          ("when" >= ${from.toISO()} AND "when" <= ${to.toISO()})
+    `
+
     const op = caseInsensitive ? 'ILIKE' : 'LIKE'
-    if (who) sql.append(SQL([` AND who_id ${op} `])).append(SQL`${exactMatch ? who : '%' + who + '%'}`)
-    if (what) sql.append(SQL([` AND what_id ${op} `])).append(SQL`${exactMatch ? what : '%' + what + '%'}`)
-    if (subject) sql.append(SQL([` AND subject_id ${op} `])).append(SQL`${exactMatch ? subject : '%' + subject + '%'}`)
+    const filterClause = SQL``
+
+    if (who) filterClause.append(SQL([` AND who_id ${op} `]).append(SQL`${exactMatch ? who : '%' + who + '%'}`))
+    if (what) filterClause.append(SQL([` AND what_id ${op} `]).append(SQL`${exactMatch ? what : '%' + what + '%'}`))
+    if (subject) filterClause.append(SQL([` AND subject_id ${op} `]).append(SQL`${exactMatch ? subject : '%' + subject + '%'}`))
+
+    sql.append(filterClause)
+    sqlCount.append(filterClause)
 
     const footer = ` ORDER BY ${sortKey} ${sortAsc ? 'ASC' : 'DESC'} LIMIT ${pageSize} OFFSET ${(page - 1) * pageSize}`
     sql.append(SQL([footer]))
 
-    const res = await this.performDatabaseOperations(sql)
-    return res.rows.map(convertToTrail)
+    const result = await this.performDatabaseOperations(async client => {
+      const count = await client.query(sqlCount)
+      const results = await client.query(sql)
```

Comment:
have you considered sending these two queries in parallel instead? Based on what I read [here](https://github.com/brianc/node-postgres/issues/516) the pg library will serialize them anyway but at least we leave that job to it instead of us serializing them. So:

```js
return Promise.all([query1, query2])
```

Reply:
from the link provided, it also seems that it may not run in the same transaction boundary, if we do it in parallel . is that understanding correct  ?

Reply:
no, I don't think it is correct understanding. we should parallelize the execution on our side and leave to the pg library to do serialization. the transactions will work regardless

Reply:
Does this seem right ?
```
const res = await this.performDatabaseOperations(async client => {
      const count = client.query(sqlCount)
      const records = client.query(sql)
      const results = await Promise.all([count, records])
      return results
    })

    return { count: res[0].rows[0].count, data: res[1].rows.map(convertToTrail) }
```

Reply:
from a functional standpoint yes, but you can simplify it to

```js
const res = await this.performDatabaseOperations(client => 
  Promise.all([
    client.query(sqlCount), 
    client.query(sql)
  ])
)
```